### PR TITLE
Added trefwoorden

### PR DESCRIPTION
**Pull request by ivo-ritense**

### CODE REVIEW

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/ZgwDocumentTrefwoordAutoConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/ZgwDocumentTrefwoordAutoConfiguration.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/ZgwDocumentTrefwoordAutoConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.springframework.boot.autoconfigure.AutoConfiguration
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
+import org.springframework.boot.autoconfigure.domain.EntityScan
+import org.springframework.context.annotation.Bean
+
+@AutoConfiguration
+@EntityScan(basePackages = ["com.ritense.documentenapi.domain"])
+class ZgwDocumentTrefwoordAutoConfiguration {
+
+    @Bean
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/ZgwDocumentTrefwoordAutoConfiguration.kt:38]: Shouldn't we annotate all beans with `ConditionalOnMissingBean` considering the discussion about opening up Valtimo for more customisation?

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordChangeset.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordChangeset.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordChangeset.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ritense.case.deployment
+
+import com.fasterxml.jackson.annotation.JsonProperty
+
+class ZgwDocumentTrefwoordChangeset(
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordChangeset.kt]:
```suggestion
data class ZgwDocumentTrefwoordChangeset(
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordChangeset.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    val changesetId: String,
+    @JsonProperty("case-definitions")
+    val caseDefinitions: List<CaseDefinitionTrefwoordCollection>
+)
+
+class CaseDefinitionTrefwoordCollection(
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordChangeset.kt]:
```suggestion
data class CaseDefinitionTrefwoordCollection(
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordDeploymentService.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/deployment/ZgwDocumentTrefwoordDeploymentService.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/domain/ZgwDocumentTrefwoord.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/domain/ZgwDocumentTrefwoord.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/domain/ZgwDocumentTrefwoordId.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package com.ritense.documentenapi.domain
+
+import jakarta.persistence.Embeddable
+import java.io.Serializable
+
+@Embeddable
+data class ZgwDocumentTrefwoordId(
+    private val value: String,
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/domain/ZgwDocumentTrefwoordId.kt]: Why are these properties private?

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/repository/ZgwDocumentTrefwoordRepository.kt [hunk 9]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/repository/ZgwDocumentTrefwoordRepository.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/domain/ZgwDocumentTrefwoordId.kt [hunk 10]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/domain/ZgwDocumentTrefwoordId.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordExporter.kt [hunk 11]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordExporter.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordImporter.kt [hunk 12]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordImporter.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordService.kt [hunk 13]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordService.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordService.kt [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+) {
+    fun getTrefwoorden(caseDefinitionName: String, pageable: Pageable): Page<ZgwDocumentTrefwoord> {
+        return zgwDocumentTrefwoordRepository.findAllByCaseDefinitionName(caseDefinitionName, pageable)
+    }
+
+    fun createTrefwoord(caseDefinitionName: String, trefwoord: String) {
+        val existingTrefwoord = zgwDocumentTrefwoordRepository.findAllByCaseDefinitionNameAndValue(caseDefinitionName, trefwoord)
+        if (existingTrefwoord != null) {
+            throw IllegalArgumentException("Trefwoord $trefwoord already exists for case definition $caseDefinitionName")
+        }
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordService.kt]:
```suggestion
        require(existingTrefwoord == null) {
            "Trefwoord $trefwoord already exists for case definition $caseDefinitionName"
        }
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/web/rest/ZgwDocumentTrefwoordResource.kt [hunk 15]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/web/rest/ZgwDocumentTrefwoordResource.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/main/resources/config/liquibase/changelog/20240222-add-documenten-api-trefwoorden.xml [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        <createTable tableName="zgw_document_trefwoord">
+            <column name="case_definition_name" type="varchar(256)">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="value" type="varchar(256)">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+        </createTable>
+    </changeSet>
+
```
Comment (theo-ritense) [inline zgw/documenten-api/src/main/resources/config/liquibase/changelog/20240222-add-documenten-api-trefwoorden.xml:33]: Shouldn't we add a foreign key to `case_definition_name` ?

File: zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordExporterIntTest.kt [hunk 17]
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2015-2023 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordExporterIntTest.kt]:
```suggestion
 * Copyright 2015-2024 Ritense BV, the Netherlands.
```

File: zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordServiceTest.kt [hunk 18]
```diff
@@ -0,0 +1,73 @@
+package com.ritense.documentenapi.service
```
Comment (theo-ritense) [inline zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/service/ZgwDocumentTrefwoordServiceTest.kt:17]:
```suggestion
/*
 * Copyright 2015-2024 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ritense.documentenapi.service
```

File: zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/web/rest/ZgwDocumentTrefwoordResourceIT.kt [hunk 19]
```diff
@@ -0,0 +1,81 @@
+package com.ritense.documentenapi.web.rest
```
Comment (theo-ritense) [inline zgw/documenten-api/src/test/kotlin/com/ritense/documentenapi/web/rest/ZgwDocumentTrefwoordResourceIT.kt:17]:
```suggestion
/*
 * Copyright 2015-2024 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ritense.documentenapi.web.rest
```

File: zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/ZgwDocumentTrefwoordAutoConfiguration.kt (hunk 2 above)
Comment (ivo-ritense) [reply zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/ZgwDocumentTrefwoordAutoConfiguration.kt:38]: Not sure a decision was made on this, but i changed it

**Review by theo-ritense — APPROVED** [review]

**Pull request merged**